modules/acton/src/restate/services/WebLeadVirtualObject.ts

```diff
@@ -0,0 +1,194 @@
+			errorCode: 400,
+		});
+	}
+	if ((await ctx.get("Status")) === "ERROR") {
+		ctx.clearAll();
+	}
+	try {
+		await SyncWithDB(ctx, "RECEIVE");
+	} catch (e) {
+		await ctx.update((_) => ({
+			Status: "ERROR",
+			Error: serializeError(e),
+		}));
+		throw e;
+	}
+
+	const status = (await ctx.get("Status")) ?? "NONEXISTANT";
+	const canRun = allowedStates.includes(status);
+	if (!canRun) {
+		throw new restate.TerminalError(
+			`Lead ID '${ctx.key}' doesn't exist or is in incorrect status '${status}'`,
+			{ errorCode: 409 },
+		);
+	}
+}
+
+export const LeadVirtualObject = restate.object({
+	name: "WebLead",
+	handlers: {
+		status: restate.handlers.object.shared(
+			async (
+				ctx: restate.ObjectSharedContext<LeadState>,
+				req: Record<string, any>,
+			): Promise<LeadState> => {
+				//await CheckAuthorization(ctx);
+				const state = await ctx.getAll();
+				if (state.Status == null) {
+					return { Status: "NONEXISTANT" };
+				}
+				console.log("state", state);
+				return state;
+			},
+		),
+		/**
+		 * Creates a new lead in the database
+		 */
+		create: restate.handlers.object.exclusive(
+			async (
+				ctx: restate.ObjectContext<LeadState>,
+				req: Record<string, any>,
+			): Promise<LeadState> => {
+				try {
+					await ctx.update((_) => ({
+						Status: "VALIDATING",
+						Request: req ?? {},
+					}));
+					// Validate the submitted lead
+					console.log("E2E test", req);
+					const parsedRequest = WebFormLeadCreateRequestSchema.safeParse(req);
+					console.log("E2E test", parsedRequest);
+					if (!parsedRequest?.success) {
+						const formattedError = fromError(parsedRequest.error);
+						throw new restate.TerminalError(
+							`Request could not be parsed - ${formattedError.message}`,
+							{
+								errorCode: 400,
+							},
+						);
+					}
+					const Lead = parsedRequest.data as WebFormLead;
+					const currentDate = new Date(await ctx.date.now());
+					await ctx.update((_) => ({
+						LeadId: ctx.key as UUID,
+						Status: "ACTIVE",
+						SchemaVersion: "1.0.0",
+						DateSubmitted: currentDate.toISOString(),
+						Integrations: {},
+						Lead: Lead,
+						UniversalRetailerId: ctx.key as UUID,
+					}));
+					// Mark the lead as ACTIVE and sync with the database
+					ctx.set<LeadState["Status"]>("Status", "ACTIVE");
+					await SyncWithDB(ctx, "SEND");
+					ctx.console.log(`Created new lead with LeadID: '${ctx.key}'`);
+				} catch (e) {
+					console.log(e);
+				}
+				// Return the status of the lead
+				return await ctx
+					.objectClient(LeadVirtualObject, ctx.key)
+					.status({ API_KEY: process.env.INTERNAL_API_TOKEN });
+			},
+		),
+		sync: restate.handlers.object.exclusive(
+			async (
+				ctx: restate.ObjectContext<LeadState>,
+				req: Record<string, any>,
+			): Promise<LeadState> => {
+				await setup(ctx, [
+					"ACTIVE",
+					"SYNCING",
+					"CLOSED",
+					"ERROR",
+					"NONEXISTANT",
+					"VALIDATING",
```

Comment:
Only allow sync method to run when the lead is in these two states
```suggestion
					"ACTIVE",
					"SYNCING",
```